src/lib/storage/fixed_string_dictionary_segment/fixed_string_vector_iterator.hpp

```diff
@@ -61,8 +61,9 @@ class FixedStringIterator : public boost::iterator_facade<FixedStringIterator<On
   }
 
   template <bool OnConstStorageLocal = OnConstStorage>
-  std::enable_if_t<OnConstStorageLocal, const std::string_view> dereference() const {  // NOLINT
-    return std::string_view{&_chars[_pos], strnlen(&_chars[_pos], _string_length)};
+  std::enable_if_t<OnConstStorageLocal, const std::string_view&> dereference() const {  // NOLINT
+    _temporary_return_value = std::string_view{&_chars[_pos], strnlen(&_chars[_pos], _string_length)};
+    return _temporary_return_value;
```

Comment:
Why is this necessary?

Reply:
To be able to return a reference.

Reply:
The previous iterator returned the string_view by value and that disabled range-based loops.

Reply:
Do you have performance numbers for this?

I have tried this approach (storing temporaries and returning references) a couple of times and always ran into problems.

Reply:
I don't have numbers only on this. There would need to be quite some rewrite I fear to have a good comparison.

Reply:
In that case and unless it is necessary for this PR, I'd leave it out. My experience is that holding temporaries in the iterators and returning references to those is a Pandora's box.

Reply:
It is necessary for this PR since operators use range-based for-loops with references.

Reply:
Mind to give me a line reference?

As far as I understand it, this violates [[iterator.concept.forward]](https://eel.is/c++draft/iterator.concept.forward):
> Pointers and references obtained from a forward iterator into a range [i, s) shall remain valid while [i, s) continues to denote a range.

Reply:
It's here:
```
../src/lib/operators/table_scan/column_like_table_scan_impl.cpp:124:10: note: use non-reference type 'std::__1::basic_string_view<char, std::__1::char_traits<char> >'
    for (const auto& value : dictionary) {
```

We could iteratively build up a vector of `std::string_view` within the iterator but I have no idea whether this would help here.

I tried to read something about such proxy iterators but could not really find a proper source. I found a paper that discusses such proxy objects (https://sean-parent.stlab.cc/papers/2008-03-sac/p171-pirkelbauer.pdf), but I don't understand a word. :D

Reply:
Another link that might help here: https://stackoverflow.com/questions/55304218/why-does-the-parallel-for-each-require-forward-iterators.

Reply:
Isn't the error just a clang warning that we could suppress locally?

Reply:
Trying it out. Did not expect that it's ok to "overrule" the request for a reference.

Reply:
`Wrange-loop-analysis` is just a performance warning, which does not seem to understand what we are doing here.

Reply:
But doesn't the warning `loop variable A is always a copy because the range of type B does not return a reference` indicate that we are now creating copies? I don't understand why there is no copy elision.

Reply:
Could you give me the entire error message please?

Reply:
Wait a sec

Reply:
Clang sees that you are using `const auto&` for the loop variable (which is correct for regular dictionaries):

```
// Warn when the loop variable is a const reference that creates a copy.
// Suggest using the non-reference type for copies.  If a copy can be prevented
// suggest the const reference type that would do so.
```

As `operator*` does not return a reference (SemaStmt.cpp:2417), clang issues a warning:

```
    // The range always returns a copy, so a temporary is always created.
    // Suggest removing the reference from the loop variable.
    SemaRef.Diag(VD->getLocation(), diag::warn_for_range_variable_always_copy)
        << VD << RangeInitType;
```

Normally, when the iterator returns a proxy object, `for (auto x : y)` would be correct and should not raise a warning for clang. In this case, we want `const auto&` for DictionarySegment and `auto` for FixedStringDictionarySegment. `const auto&` avoids copies for DictionarySegments and works for `FixedStringDictionarySegment` because of reference lifetime extension.

I believe that clang's error message is a bit misleading. This is not really a copy being made and lifetime-extended, but a proxy.

Reply:
This is ref vs no-ref in the iterator:

```diff
 +----------------+--------------------+-------+--------------------+-------+------------+---------+
 | Benchmark      | prev. iter/s       | runs  | new iter/s         | runs  | change [%] | p-value |
 +----------------+--------------------+-------+--------------------+-------+------------+---------+
 | TPC-H 01       | 0.6443614363670349 | 39    | 0.6555484533309937 | 40    | +2%        |  0.0119 |
 | TPC-H 02       | 72.77520751953125  | 4367  | 69.72757720947266  | 4184  | -4%        |  0.0000 |
 | TPC-H 03       | 7.245577335357666  | 435   | 7.494514465332031  | 450   | +3%        |  0.0000 |
 | TPC-H 04       | 6.830118656158447  | 410   | 6.830428600311279  | 410   | +0%        |  0.9915 |
-| TPC-H 05       | 3.8460309505462646 | 231   | 3.63138484954834   | 218   | -6%        |  0.0000 |
-| TPC-H 06       | 226.85472106933594 | 13612 | 206.52418518066406 | 12392 | -9%        |  0.0000 |
 | TPC-H 07       | 7.674581527709961  | 461   | 7.545098304748535  | 453   | -2%        |  0.0097 |
 | TPC-H 08       | 6.617753505706787  | 398   | 6.371017932891846  | 383   | -4%        |  0.0000 |
 | TPC-H 09       | 1.5927869081497192 | 96    | 1.577647089958191  | 95    | -1%        |  0.0909 |
 | TPC-H 10       | 3.0369393825531006 | 183   | 2.998673677444458  | 180   | -1%        |  0.0466 |
 | TPC-H 11       | 30.272178649902344 | 1817  | 30.27552604675293  | 1817  | +0%        |  0.9270 |
 | TPC-H 12       | 12.562972068786621 | 754   | 12.499527931213379 | 750   | -1%        |  0.0000 |
 | TPC-H 13       | 2.7039954662323    | 163   | 2.703594207763672  | 163   | -0%        |  0.9584 |
 | TPC-H 14       | 46.71722412109375  | 2804  | 45.88617706298828  | 2754  | -2%        |  0.0000 |
-| TPC-H 15       | 97.61774444580078  | 5858  | 92.44818115234375  | 5547  | -5%        |  0.0000 |
 | TPC-H 16       | 7.363173007965088  | 442   | 7.072004795074463  | 425   | -4%        |  0.0000 |
-| TPC-H 17       | 4.964298248291016  | 299   | 4.561469554901123  | 274   | -8%        |  0.0000 |
 | TPC-H 18       | 1.2641243934631348 | 76    | 1.2041053771972656 | 73    | -5%        |  0.0010 |
-| TPC-H 19       | 9.576200485229492  | 575   | 8.67216968536377   | 521   | -9%        |  0.0000 |
 | TPC-H 20       | 28.287826538085938 | 1698  | 27.54608917236328  | 1653  | -3%        |  0.0000 |
+| TPC-H 21       | 1.1576995849609375 | 70    | 1.2195662260055542 | 74    | +5%        |  0.0000 |
-| TPC-H 22       | 16.134994506835938 | 969   | 14.36899471282959  | 863   | -11%       |  0.0000 |
 | geometric mean |                    |       |                    |       | -3%        |         |
 +----------------+--------------------+-------+--------------------+-------+------------+---------+
```

Reply:
And that's for FixedStringDictionaries, right?

Reply:
Yes, all strings are fixed string dictionary segments. "prev" is the version with references, "new" is the version without.

Any ideas what is happening here? Could it be that there are actually copies being created?

Reply:
Mmmh, I don't think that it is in the like_matcher, as that is only used for a couple of queries. My guess is on the scan of dictionary-compressed segments, and more specifically somewhere around `ColumnVsValueTableScanImpl::_get_search_value_id`.

Just to show that the references are broken:

```
TEST_F(FixedStringVectorTest, Foo) {
  std::vector<pmr_string> v = {"abc", "x", "ghi"};
  auto fs_vector = FixedStringVector{v.cbegin(), v.cend(), 3};


  const auto& fs_lower_bound = *std::lower_bound(fs_vector.cbegin(), fs_vector.cend(), "def");
  const auto& v_lower_bound = *std::lower_bound(v.cbegin(), v.cend(), "def");
  EXPECT_EQ(fs_lower_bound, v_lower_bound);
}
```

```
[ RUN      ] FixedStringVectorTest.Foo
../src/test/storage/fixed_string_vector_test.cpp:76: Failure
Expected equality of these values:
  fs_lower_bound
    Which is: { ' ' (32, 0x20), '\xE9' (233), '\xE3' (227), '\xEF' (239), '\xFC' (252), '\x7F' (127), '\0', '\0', '\xF1' (241), '\x82' (130), 'N' (78, 0x4E), 'p' (112, 0x70), '\x13' (19), 'V' (86, 0x56), '\0', '\0', '\0', 'B' (66, 0x42) }
  v_lower_bound
    Which is: { 'x' (120, 0x78) }
[  FAILED  ] FixedStringVectorTest.Foo (0 ms)
```